risc0/zkvm/src/host/server/exec/executor.rs

```diff
@@ -100,6 +101,39 @@ impl OpCodeResult {
     }
 }
 
+/// Error variants used in the Executor
+pub enum ExecutorError {
+    /// This variant represents an instance of Session that Faulted
+    Fault(Session),
+    /// This variant represents all other errors
+    Error(anyhow::Error),
+}
+
+use std::error::Error as StdError;
+unsafe impl Sync for ExecutorError {}
+unsafe impl Send for ExecutorError {}
+
+impl std::fmt::Debug for ExecutorError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        fmt::Display::fmt(&self, f)
+    }
+}
+
+impl std::fmt::Display for ExecutorError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            ExecutorError::Error(e) => write!(f, "{e}"),
+            ExecutorError::Fault(s) => write!(
+                f,
+                "Faulted Session [Image ID: {}]",
+                s.segments[0].resolve().unwrap().pre_image.compute_id()
```

Comment:
We probably should not call `resolve` in this context since it could result in a network call that may fail. Since we are already in an error context in this code, we should be careful to avoid inducing another error.

Reply:
good catch!